semantics-3.0/Semantics.hs

```diff
@@ -236,206 +227,219 @@ data ReduceWarning = ReduceNoWarning
+        amountToPay = evalValue env state val
+        moneyToPay  = Lovelace amountToPay
+        in  if amountToPay <= 0
+            then Reduced (ReduceNonPositivePay accId payee moneyToPay) Nothing state cont
+            else let
+                (paidMoney, newAccs) = withdrawMoneyFromAccount accId moneyToPay (accounts state)
+                warning = if paidMoney < moneyToPay
+                          then ReducePartialPay accId payee paidMoney moneyToPay
                           else ReduceNoWarning
                 (payEffect, finalAccs) = giveMoney payee paidMoney newAccs
-            in Reduced warning payEffect (state { accounts = finalAccs }) cont
-      where amountToPay = evalValue env state val
-    If obs cont1 cont2 -> Reduced ReduceNoWarning ReduceNoEffect state cont
-      where cont = if evalObservation env state obs then cont1 else cont2
-    When _ timeout cont
+                in Reduced warning payEffect (state { accounts = finalAccs }) cont
+
+    If obs cont1 cont2 -> let
+        cont = if evalObservation env state obs then cont1 else cont2
+        in Reduced ReduceNoWarning Nothing state cont
+
+    When _ timeout cont -> let
+        startSlot = ivFrom (slotInterval env)
+        endSlot   = ivTo (slotInterval env)
         -- if timeout in future – do not reduce
-        | endSlot < timeout -> NotReduced
+        in if endSlot < timeout then NotReduced
         -- if timeout in the past – reduce to timeout continuation
-        | timeout <= startSlot -> Reduced ReduceNoWarning ReduceNoEffect state cont
+        else if timeout <= startSlot then Reduced ReduceNoWarning Nothing state cont
         -- if timeout in the slot range – issue an ambiguity error
-        | otherwise -> ReduceError ReduceAmbiguousSlotInterval
-      where
-        startSlot = fst $ slotInterval env
-        endSlot   = snd $ slotInterval env
-    Let valId val cont -> Reduced warn ReduceNoEffect newState cont
-      where
-        evVal = evalValue env state val
+        else AmbiguousSlotIntervalError
+
+    Let valId val cont -> let
+        evaluatedValue = evalValue env state val
         boundVals = boundValues state
-        newState = state { boundValues = Map.insert valId evVal boundVals }
+        newState = state { boundValues = Map.insert valId evaluatedValue boundVals }
         warn = case Map.lookup valId boundVals of
-              Just oldVal -> ReduceShadowing valId oldVal evVal
+              Just oldVal -> ReduceShadowing valId oldVal evaluatedValue
               Nothing -> ReduceNoWarning
+        in Reduced warn Nothing newState cont
 
 
--- REDUCE ALL
-
-data ReduceAllResult = ReducedAll [ReduceWarning] [ReduceEffect] State Contract
-                     | ReduceAllError ReduceError
-  deriving (Eq,Ord,Show,Read)
+data ReduceAllResult = ReduceAllSuccess [ReduceWarning] [Payment] State Contract
+                     | ReduceAllAmbiguousSlotIntervalError
+  deriving (Eq,Ord,Show)
```

Comment:
We should probably replace "ReduceAll" with "ReduceContractUntilQuiescent" or some abbreviation of it (everywhere).

Reply:
What do you think about `QuiescenceResult`?